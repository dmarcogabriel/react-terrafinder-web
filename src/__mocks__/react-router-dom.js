const originalModule = jest.requireActual('react-router-dom');

module.exports = {
  ...originalModule,
  useHistory: () => ({
    history: { push: jest.fn() },
  }),
};
