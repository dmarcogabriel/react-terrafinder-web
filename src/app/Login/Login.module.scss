@import 'styles';

$padding: 1.5rem;

.container {
  display: flex;
  flex-direction: column;

  @media ($desktop) {
    flex-direction: row;
  }

  .welcome {
    background-image: url('./static/background-mobile.jpeg');
    background-size: cover;
    min-height: 10rem;
    background-repeat: no-repeat;
    padding: $padding;

    @media ($desktop) {
      background-image: url('./static/background.jpg');
      height: 100vh;
      width: 50%;
      background-size: cover;
    }

    > h1 {
      font-size: 5rem;
      color: $white;

      @media ($desktop) {
        color: $gray-light;
        margin: 2rem 3rem;
      }
    }
  }

  .login {
    padding: $padding;

    @media ($desktop) {
      width: 50%;
    }

    .form {
      @media ($desktop) {
        padding: 2rem 2.5rem;
      }

      > h2 {
        color: $primary-dark;
        font-size: 2rem;
        margin-bottom: 1rem;
        text-align: center;
      }

      > p {
        color: $gray-light;
        text-align: center;
      }

      .my {
        margin: 1.5rem 0;
      }

      .my2 {
        margin: 2.5rem 0;
      }

      .groupedInputs {
        @media ($desktop) {
          display: flex;
          justify-content: space-between;
          align-items: flex-end;
        }
      }

      .actionButton {
        background: $green;
        width: 100%;
        text-align: center;
        margin: 2rem 0 2.5rem;
      }
    }
  }
}
