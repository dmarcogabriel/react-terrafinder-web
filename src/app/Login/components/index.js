import { LoginContainer } from './LoginContainer';
import { LoginH2, LoginSubtitle } from './LoginTypography';

export { LoginContainer, LoginH2, LoginSubtitle };
