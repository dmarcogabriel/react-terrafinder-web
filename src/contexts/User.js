import React, { createContext, useState, useEffect } from 'react';
import { USER_KEY } from 'constants/userKey';
import api from 'services/api';

export const Context = createContext();

export const UserProvider = ({ children, storedUser }) => {
  const [currentUser, setCurrentUser] = useState(null);
  const [loadingUser, setLoadingUser] = useState();

  const login = ({ token, email, firstName, _id, ...user }) => {
    const data = { _id, firstName, email, token };

    window.localStorage.setItem(USER_KEY, JSON.stringify(data));
    setCurrentUser({ ...user, ...data });
  };

  const userIsLogged = () => !!currentUser;

  const logout = () => {
    window.localStorage.removeItem(USER_KEY);
    setCurrentUser(null);
    window.location.href = '/home';
  };

  const loadUserData = async () => {
    setLoadingUser(true);

    try {
      const { data: res } = await api.get(`users/${storedUser._id}`);
      console.log(res.data);
      setCurrentUser({ ...res.data.user, token: storedUser.token });
      setLoadingUser(false);
    } catch (err) {
      logout();
      setLoadingUser(false);
    }
  };

  const updateUser = (data) => {
    setCurrentUser((old) => ({ ...old, ...data }));
  };

  useEffect(() => {
    if (storedUser) loadUserData();
  }, []);

  return (
    <Context.Provider
      value={{ currentUser, login, logout, updateUser, userIsLogged }}
    >
      {!loadingUser && children}
    </Context.Provider>
  );
};
