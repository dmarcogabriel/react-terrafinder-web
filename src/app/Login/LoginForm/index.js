import React from 'react';
import { Link } from 'react-router-dom';
import Input from 'common/Input';
import Button from 'common/Button';
import { useFormik } from 'formik';
import { object, string } from 'yup';
import api from 'services/api';
import { useUser } from 'hooks/useUser';
import { form, actionButton, my, my2 } from '../Login.module.scss';

export default function LoginForm() {
  const { login } = useUser();

  const { values, handleChange, errors, handleSubmit } = useFormik({
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema: object().shape({
      email: string().email('Deve ser e-mail').required('Campo obrigatório'),
      password: string()
        .min(6, 'Mínimo 6 caracteres')
        .required('Campo Obrigatório'),
    }),
    async onSubmit() {
      const response = await api.post('login', values);

      const userData = await api.get(`users/${response.data.userId}`, {
        headers: {
          'x-access-token': response.data.token,
        },
      });

      login({ ...userData.data.user, token: response.data.token });
    },
  });

  return (
    <div className={form}>
      <h2>Faça login para acessar seu painel</h2>

      <p>
        Não tem conta ainda?
        <Link to="/login/register"> Cadastre-se aqui</Link>
      </p>

      <div className={my2}>
        <Input
          label="E-mail"
          value={values.email}
          onChange={handleChange('email')}
          errorMessage={errors.email}
        />

        <Input
          label="Senha"
          type="password"
          value={values.password}
          onChange={handleChange('password')}
          errorMessage={errors.password}
        />

        <p className={my}>
          Esqueceu sua senha?
          <Link to="/login/forgot-password">Clique aqui para redefinir</Link>
        </p>

        <Button className={actionButton} onClick={handleSubmit}>
          Fazer Login
        </Button>
      </div>
    </div>
  );
}
