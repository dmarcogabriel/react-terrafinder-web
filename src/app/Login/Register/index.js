import React from 'react';
import { Link } from 'react-router-dom';
import Button from 'common/Button';
import Input from 'common/Input';
import { useFormik } from 'formik';
import { object, string } from 'yup';
import { form, groupedInputs, actionButton } from '../Login.module.scss';

export default function Register() {
  const { values, errors, handleChange, handleSubmit } = useFormik({
    initialValues: {
      firstName: '',
      lastName: '',
      email: '',
      cpf: '',
      phone: '',
      password: '',
    },
    validationSchema: object().shape({
      firstName: string().required('Campo obrigatório!'),
      lastName: string().required('Campo obrigatório!'),
      email: string().email().required('Campo obrigatório!'),
      cpf: string().required('Campo obrigatório!'),
      phone: string().required('Campo obrigatório!'),
      password: string()
        .min(6, 'Mínimo 6 caracteres')
        .required('Campo obrigatório!'),
    }),
    onSubmit() {
      console.log(values);
    },
  });
  return (
    <div className={form}>
      <h2>Cadastre-se para cadastrar seu anúncio</h2>

      <p>
        Já tem uma conta?
        <Link to="/login"> Faça o login aqui</Link>
      </p>

      <div className={groupedInputs}>
        <Input
          label="Primeiro nome"
          value={values.firstName}
          onChange={handleChange('firstName')}
          errorMessage={errors.firstName}
        />

        <Input
          label="Sobrenome"
          value={values.lastName}
          onChange={handleChange('lastName')}
          errorMessage={errors.lastName}
        />
      </div>

      <Input
        label="E-mail"
        type="email"
        value={values.email}
        onChange={handleChange('email')}
        errorMessage={errors.email}
      />

      <div className={groupedInputs}>
        <Input
          label="Telefone para contato"
          value={values.phone}
          onChange={handleChange('phone')}
          errorMessage={errors.phone}
        />

        <Input
          label="CPF"
          value={values.cpf}
          onChange={handleChange('cpf')}
          errorMessage={errors.cpf}
        />
      </div>

      <Input
        label="Senha"
        type="password"
        value={values.password}
        errorMessage={errors.password}
        onChange={handleChange('password')}
      />

      <Button className={actionButton} onClick={handleSubmit}>
        Cadastrar
      </Button>

      <p>
        Se cadastrando, você concorda com nossos
        <Link to="/">Termos de Uso</Link>
        {' e '}
        <Link to="/">Política de privacidade</Link>
      </p>
    </div>
  );
}
