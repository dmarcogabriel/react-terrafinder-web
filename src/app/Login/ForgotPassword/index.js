import React from 'react';
import { Button, Input } from 'common/components';
import { form, my2, actionButton } from '../Login.module.scss';

export default function ForgotPassword() {
  return (
    <div className={form}>
      <h2>Fa√ßa login para acessar seu painel</h2>

      <p>Insira seu e-mail abaixo para redefinir a senha</p>

      <div className={my2}>
        <Input label="E-mail" onChange={console.log} />

        <Button className={actionButton}>Redefinir Senha</Button>
      </div>
    </div>
  );
}
