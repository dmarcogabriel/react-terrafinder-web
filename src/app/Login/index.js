import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { container, welcome, login } from './Login.module.scss';
import LoginForm from './LoginForm';
import ForgotPassword from './ForgotPassword';
import Register from './Register';

export default function Login() {
  return (
    <div className={container}>
      <div className={welcome}>
        <h1>Bem vindo!</h1>
      </div>

      <div className={login}>
        <Switch>
          <Route exact path="/login">
            <LoginForm />
          </Route>

          <Route path="login/register">
            <Register />
          </Route>

          <Route path="login/forgot-password">
            <ForgotPassword />
          </Route>
        </Switch>
      </div>
    </div>
  );
}
