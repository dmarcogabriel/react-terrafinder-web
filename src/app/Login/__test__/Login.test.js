import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { UserProvider } from 'contexts/User';
import Login from '..';

const Comp = () => (
  <UserProvider>
    <BrowserRouter>
      <Login />
    </BrowserRouter>
  </UserProvider>
);

describe('<Login />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Comp />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should replace page', () => {
    jest.doMock('hooks/useUser', () => ({
      useUser() {
        return { currentUser: true };
      },
    }));

    render(<Comp />);
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<Comp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
