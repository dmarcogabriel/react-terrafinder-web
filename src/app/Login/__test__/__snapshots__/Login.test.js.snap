// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> matches snapshot 1`] = `
<div>
  <div>
    <h1>
      Bem vindo!
    </h1>
  </div>
  <div />
</div>
`;
